File: loyaltyX_backend/models/Card.js
import { DataTypes } from 'sequelize'
import sequelize from '../db.js'
import User from './User.js'

const Card = sequelize.define('Card', {
  id: {
    type: DataTypes.STRING,
    allowNull: false,
    unique: true,
    primaryKey: true
  },
  userId: {
    type: DataTypes.STRING,
    allowNull: false,
    references: {
        model: User,
        key: 'userId',
    },
    onUpdate: 'CASCADE',
    onDelete: 'CASCADE'
  },
  cardHolderName: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  cardNumber: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  cardYear: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  cardMonth: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  cardCvv: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  default: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  }
}, {
  tableName: 'cards',
})

export default Card

File: loyalty/lib/models/card.dart
class Card {
  Card({
    required this.id,
    required this.userId,
    required this.cardHolderName,
    required this.cardNumber,
    required this.cardYear,
    required this.cardMonth,
    required this.cardCvv,
    this.isDefault = false,
  });

  final String id;
  final String userId;
  final String cardHolderName;
  final String cardNumber;
  final int cardYear;
  final int cardMonth;
  final String cardCvv;
  bool isDefault;

  static Card fromJson(Map<String, dynamic> data) {
    return Card(
      id: data['id'],
      cardHolderName: data['cardHolderName'],
      userId: data['userId'],
      cardNumber: data['cardNumber'],
      cardYear: data['cardYear'],
      cardMonth: data['cardMonth'],
      cardCvv: data['cardCvv'],
      isDefault: data['default'] ?? false,
    );
  }

  static Map<String, dynamic> toJson(Card card) {
    return {
      'id': card.id,
      'userId': card.userId,
      'cardHolderName': card.cardHolderName,
      'cardNumber': card.cardNumber,
      'cardYear': card.cardYear,
      'cardMonth': card.cardMonth,
      'cardCvv': card.cardCvv,
      'default': card.isDefault,
    };
  }
}
